WIP - Added prompt user for url

jklepatch opened the pull request:
This is not ready yet, I would just like to have some feedback.

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #17047 
- [ ] These changes do not require tests because _____

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/17062)
<!-- Reviewable:end -->

jklepatch committed:
# This is a combination of 2 commits.
# The first commit's message is:

Added prompt user for url

Fix Cargo.lock corrupted problem

Refactored event handler for CTRL + L

Made sanitize_and_parse_url() handle correctly parse errors

# This is the commit message #2:

Made the prompt default to the current url

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @pcwalton (or someone else) soon.

paulrouget commented on ports/glutin/window.rs:
```diff
@@ -1268,12 +1269,33 @@ impl WindowMethods for Window {
                     self.event_queue.borrow_mut().push(WindowEvent::Reload);
                 }
             }
+            //On windows 8.1 when pressing ctrl + anything, ch = Some('\u{unicode_value}')
```
We need to investigate that before merging this PR.

paulrouget reviewed (commented):
It's weird. The changes in Cargo.lock make it impossible for me to compile.

```
./mach build -d --verbose
cargo build -v
error: failed to parse lock file at: /Users/paul/git/servo/Cargo.lock

Caused by:
  package `gcc 0.3.45 (registry+https://github.com/rust-lang/crates.io-index)` is specified as a dependency, but is missing from the package list
```

I'll see later what's going on.

paulrouget commented on ports/glutin/window.rs:
```diff
@@ -1402,6 +1424,10 @@ fn is_printable(key_code: VirtualKeyCode) -> bool {
 }
 
+fn sanitize_url(url: String) -> Result<ServoUrl, ()> {
+    Ok(ServoUrl::parse(url.trim()).unwrap())
```
FYI, unwrap() on a failing Result will panic. See https://is.gd/RAdLog

paulrouget commented on ports/glutin/window.rs:
```diff
@@ -1268,12 +1269,33 @@ impl WindowMethods for Window {
+                if let Some(true) = PREFS.get("shell.builtin-key-shortcuts.enabled").as_boolean() {
+                    let user_input: String;
+                    //TODO: 
+                    // 1) make it work when clicking on "validate" button - Currently only works when pressing enter
```
Maybe it has to do with the fact we're blocking the main thread.
On Mac, the buttons work, but we get the beatchball.

Maybe this code should run in a different thread.

paulrouget commented on ports/glutin/window.rs:
```diff
@@ -1402,6 +1419,10 @@ fn is_printable(key_code: VirtualKeyCode) -> bool {
     }
 }
 
+fn sanitize_and_parse_url(url: String) -> Result<ServoUrl, ()> {
```
Actually, we might want to use `ParseError` for now instead of an empty error.

paulrouget commented on ports/glutin/window.rs:
```diff
@@ -1268,12 +1269,28 @@ impl WindowMethods for Window {
+                    // 1) make it work when clicking on "validate" button - Currently only works when pressing enter
+                    // 2) Erratic behavior - sometime the dialog reloads empty after pressing enter
+                    // (windows 8.1)
+                    if let Some(user_input) = tinyfiledialogs::input_box("Enter a URL", "URL:", "")  {
```
The default value should be the current url, not `""`.

I'm wondering if we have any locales support. I'm not sure we want to keep english sentences in the code.

paulrouget commented:
@jdm is tinyfiledialogs thread safe? It appears to work well here (osx) if I open a prompt from a different thread. But I'm not sure it's the right thing to do.

jdm commented:
I do not believe the library does anything special to ensure that it works from non-main threads. I suspect that at least one platform will not work as intended if used that way.

jklepatch committed:
Added prompt user for url

Fix Cargo.lock corrupted problem

Refactored event handler for CTRL + L

Made sanitize_and_parse_url() handle correctly parse errors

Made the prompt default to the current url

Made sanitize_and_parse_url() return ParseError instead of empty error

jklepatch commented:
Currently, in `handle_key()` (window.rs) I am matching `ctrl + L` on `Key` instead of `ch`. This is a temporary solution. Once this related issue [PR](https://github.com/servo/servo/issues/17146) is solved, I can match on `ch` and the PR will be ready.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #17184) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
Closing due to inactivity.
